### DISCUSSION TITLE
Lua Keymaps

### DISCUSSION THREAD

**Discussion by kevintraver** [CopilotC-Nvim/CopilotChat.nvim#230] [Q&A] [answered]
What are the lua equivalents of the prompt keymaps described in the docs?

It looks like it should be something like this:
```
{
    "<leader>cce",
    function()
      require("CopilotChat").ask( ... )
    end,
    desc = "CopilotChat - Explain code",
}
```

What exactly should be passed into the `ask` parameter?

**Comment by deathbeam** [comment]
For ask, you can see the example api, e.g first parameter takes string which is the actual prompt and second parameter accepts config (which is what you can see in Configuration section). So it depends on what you want to achieve I guess with that mapping

**Reply by kevintraver** [reply]
What would be the parameters to create the same mapping as `:CopilotChatExplain`?

**Reply by deathbeam** [reply, marked as answer]
Ah, like that. You can do:

```lua
require("CopilotChat").ask("/COPILOT_EXPLAIN Write an explanation for the code above as paragraphs of text.")
```

and its the same

**Reply by kevintraver** [reply]
That works great, thanks!

Just curious about `:CopilotChatCommitStaged`

I think it would be something like this, but I'm not sure what to set for `source`

```
          local copilot_chat = require("CopilotChat")
          local config = require("CopilotChat.config")

          local selection = config.prompts.CommitStaged.selection(source)
          copilot_chat.ask(selection, config.prompts.CommitStaged.prompt)
```
https://github.com/CopilotC-Nvim/CopilotChat.nvim/blob/2b53687d0ccc025f07972dd8ac11cc74b9f17722/lua/CopilotChat/config.lua#L132-L137

**Reply by deathbeam** [reply]
You want to pass just the function, so the selection would be like in config, e.g

```lua
function(source) 
   return select.gitdiff(source, true) 
 end
 ```
 
 and the full call would look like this:
 
 ```lua
local copilot_chat = require("CopilotChat")
local config = require("CopilotChat.config")
copilot_chat.ask(config.prompts.CommitStaged.prompt, {
   selection = config.prompts.CommitStaged.selection
})
 ```

**Reply by kevintraver** [reply]
Thank you!

**Reply by kevintraver** [reply]
That works great. Thanks!

It might be nice to have a **lua** function for each default prompt. 

That would simplify the example above, and make it easier to use the api like this:

```lua
  require("CopilotChat").commitStaged()
````

Thoughts? Happy to help with a PR.

**Reply by kevintraver** [reply]
> You want to pass just the function, so the selection would be like in config, e.g
> 
> ```lua
> function(source) 
>    return select.gitdiff(source, true) 
>  end
> ```
> 
> and the full call would look like this:
> 
> ```lua
> local copilot_chat = require("CopilotChat")
> local config = require("CopilotChat.config")
> copilot_chat.ask(config.prompts.CommitStaged.prompt, {
>   selection = config.prompts.CommitStaged.selection
> })
> ```

Is there a way to check if anything is actually staged, and then if not notify instead of running `.ask`?

Right now this still runs the command, but then just generates an example message if nothing is staged.
